src/util/isDomainSpecifiedByUser.ts

```diff
@@ -28,3 +30,198 @@ export function isDomainSpecifiedByUser(
+ *
+ * @return [min, max] domain if it's well-formed; undefined if the domain is invalid
+ */
+export function numericalDomainSpecifiedWithoutRequiringData(
+  userDomain: AxisDomain | undefined,
+  allowDataOverflow: boolean,
+) {
+  if (!allowDataOverflow) {
+    // Cannot compute data overflow if the data is not provided
+    return undefined;
+  }
+  if (typeof userDomain === 'function') {
+    try {
+      const result = userDomain(undefined, allowDataOverflow);
+      if (isWellFormedNumberDomain(result)) {
+        return result;
+      }
+    } catch (ex) {
+      /* ignore the exception and compute domain from data later */
+    }
+  }
+  if (Array.isArray(userDomain) && userDomain.length === 2) {
+    const [providedMin, providedMax] = userDomain;
+    let finalMin, finalMax: number;
+
+    if (isWellBehavedNumber(providedMin)) {
+      finalMin = providedMin;
+    } else if (typeof providedMin === 'function') {
+      try {
+        finalMin = providedMin(undefined);
+      } catch (ex) {
+        /* ignore the exception and compute domain from data later */
+      }
+    }
+
+    if (isWellBehavedNumber(providedMax)) {
+      finalMax = providedMax;
+    } else if (typeof providedMax === 'function') {
+      try {
+        finalMax = providedMax(undefined);
+      } catch (ex) {
+        /* ignore the exception and compute domain from data later */
+      }
+    }
+
+    const candidate = [finalMin, finalMax];
+    if (isWellFormedNumberDomain(candidate)) {
+      return candidate;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * So Recharts allows users to provide their own domains,
+ * but it also places some expectations on what the domain is.
+ * We can improve on the typescript typing, but we also need a runtime test
+ * to observe that the user-provided domain is well-formed,
+ * that is: an array with exactly two numbers.
+ * If the user-provided domain is not well-formed,
+ * this function will return undefined - in which case we should traverse the data to calculate the real domain.
+ *
+ * This function is for parsing the numerical domain only.
+ *
+ * @param userDomain external prop, user provided, before validation. Can have various shapes: array, function, special magical strings inside too.
+ * @param dataDomain calculated from data. Can be undefined, as an option for performance optimization
+ * @param allowDataOverflow provided by users. If true then the data domain wins
+ * @param allowDecimals if true then recharts is allowed to generate domain with decimal numbers; if false the numbers are always integers
+ *
+ * @return [min, max] domain if it's well-formed; undefined if the domain is invalid
+ */
+export function parseNumericalUserDomain(
+  userDomain: AxisDomain | undefined,
+  dataDomain: NumberDomain | undefined,
+  allowDataOverflow: boolean,
+  allowDecimals: boolean,
+): NumberDomain | undefined {
+  if (!allowDataOverflow && dataDomain == null) {
```

Comment:
It's not the prettiest function ever but I want to focus on the selectors rather than refactoring this 🙃 it's well covered so it should be easy to rewrite later if we ever want to.